### ISSUE TITLE
ValueError: `training_neg_sample_num` should be 0 if inter_feat have label_field [label].

### ISSUE THREAD

**Issue by laxmimerit** [RUCAIBox/RecBole#674]
I am getting the following error for `LightGCN`

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-14-0b2a29229228> in <module>
     47 
     48 # dataset splitting
---> 49 train_data, valid_data, test_data = data_preparation(config, dataset)
     50 
     51 # model loading and initialization

c:\users\laxmi\downloads\recbole\recbole\data\utils.py in data_preparation(config, dataset, save)
    101     if config['training_neg_sample_num']:
    102         if dataset.label_field in dataset.inter_feat:
--> 103             raise ValueError(f'`training_neg_sample_num` should be 0 '
    104                              f'if inter_feat have label_field [{dataset.label_field}].')
    105         train_distribution = config['training_neg_sample_distribution'] or 'uniform'

ValueError: `training_neg_sample_num` should be 0 if inter_feat have label_field [label].

```
If I set `training_neg_sample_num=0` then I get another error saying
```
c:\users\laxmi\downloads\recbole\recbole\model\general_recommender\lightgcn.py in calculate_loss(self, interaction)
    136         user = interaction[self.USER_ID]
    137         pos_item = interaction[self.ITEM_ID]
--> 138         neg_item = interaction[self.NEG_ITEM_ID]
    139 
    140         user_all_embeddings, item_all_embeddings = self.forward()

c:\users\laxmi\downloads\recbole\recbole\data\interaction.py in __getitem__(self, index)
    103     def __getitem__(self, index):
    104         if isinstance(index, str):
--> 105             return self.interaction[index]
    106         else:
    107             ret = {}

KeyError: 'neg_pid'
```

Any idea, how to remove this error?

**Comment by chenyushuo** [comment]
Have you set `threshold`? It should be removed because `threshold` is used to set `label`.

If you want to remove some interactions, you can use `lowest_val` or `highest_val`.

For more information, you can refer to [this](https://recbole.io/docs/user_guide/data/data_args.html).

**Comment by laxmimerit** [comment]
Hi,
Thanks for the quick reply.
I have also tried removing `threshold` but still, I get the error
This is my parameters setting
```
parameter_dict = {
    "field_separator": "\t",
    "seq_separator": " ",
    "USER_ID_FIELD":"uid",
    "ITEM_ID_FIELD":"pid",
    "RATING_FIELD":'rating',
    "TIME_FIELD": 'timestamp',
    "load_col": None,
    "epochs": 10,
    "train_batch_size": 4096,
    "eval_batch_size": 4096,
    "learner": 'adam',
    "learning_rate": 0.001,
    "drop_filter_field" : True,
    "eval_setting": "RO_RS",
    "group_by_user": False,
    "valid_metric": "AUC",
    "metrics": ['AUC', 'LogLoss'],
    "training_neg_sample_num":0, # tried with 1 as well
    "fill_nan": True,
    "normalize_all": True,
    "split_ratio":[0.6, 0.2, 0.2]
}
```
Would you please suggest to me the right dictionary params?

**Comment by chenyushuo** [comment]
`eval_setting` should be `RO_RS,uni100`.
You can refer [this page](https://recbole.io/docs/user_guide/evaluation_support.html) to learn more about `eval_setting`.

**Comment by laxmimerit** [comment]
Thank you.

**Comment by laxmimerit** [comment]
I got `LIGHTGCN` working. Thanks for the help.
After training, I load the model and get a recommendation by using the below code

```
input_inter = Interaction({
    'uid': torch.tensor([uid_number]) #numerical value of user_id
})

input_inter = input_inter.to(device)
with torch.no_grad():
    scores = model.full_sort_predict(input_inter.cpu())
print(scores.shape)
print(scores)
```
I get results for some `uid` but don't get for all `uid`. I assume that `uid` should be in training data so I used only those `uid` that were present in training data.
```
train_data.dataset.inter_feat.interaction['uid']
-----
OUT:
tensor([646995, 271091, 387667,  ..., 173207, 102803,  37796])
```
If I pass these `uid`, it throws an error
```
<decorator-gen-60> in time(self, line, cell, local_ns)

<timed exec> in <module>()

/usr/local/lib/python3.6/dist-packages/recbole/data/interaction.py in to(self, device, selected_field)
    161         else:
    162             for k in self.interaction:
--> 163                 ret[k] = self.interaction[k].to(device)
    164         return Interaction(ret)
    165 

RuntimeError: CUDA error: device-side assert triggered
```

I have also tried passing `uid` present in the test and valid dataset but no success. I get output If I pass some smaller number like `1, 2, 3, 100, 1000, ...`. 
I don't understand why it is working for some numbers and not working for others.

Please suggest.

**Comment by chenyushuo** [comment]
Can you offer the source code of predict all uid from `train_data.dataset.inter_feat.interaction['uid']`? We will check it.

**Comment by laxmimerit** [comment]
Hi,
Sorry I did not understand. Do you mean whole code training to testing?

**Comment by chenyushuo** [comment]
Yes, the whold code.

**Comment by laxmimerit** [comment]
**Update 2**
I get output for smaller numbers and those numbers are not in user_id.
Here is the complete code
```
import numpy as np
import torch
from recbole.config import Config
from recbole.data import create_dataset, data_preparation
from recbole.data.interaction import Interaction

from recbole.quick_start import run_recbole

from logging import getLogger
from recbole.config import Config
from recbole.data import create_dataset, data_preparation
from recbole.model.general_recommender import BPR, LightGCN
from recbole.trainer import Trainer
from recbole.utils import init_seed, init_logger

parameter_dict = {
    "data_path":'/path/to/data',
    "USER_ID_FIELD":"uid",
    "ITEM_ID_FIELD":"pid",
    "RATING_FIELD":'rating',
    "TIME_FIELD": 'timestamp',
    "load_col": None,
    "epochs": 2,
    "train_batch_size": 4096,
    "eval_batch_size": 4096,
    "learner": 'adam',
    "learning_rate": 0.001,
    "eval_setting": "RO_RS,uni100",
}

if __name__ == '__main__':
    # Load config, dataset and model (if you have load these things, you can skip these codes)
    config = Config(model='LightGCN', dataset='set1', config_dict=parameter_dict)
    dataset = create_dataset(config)
    train_data, valid_data, test_data = data_preparation(config, dataset)
    model = LightGCN(config, train_data)
    checkpoint = torch.load('/path/to/model/LightGCN-Jan-14-2021_08-22-43.pth')
    model.load_state_dict(checkpoint['state_dict'])
    model.eval()
    
def get_recommendation(cid, model):
  device = "cuda:0"
  model = model.to(device)

  input_inter = Interaction({
      'customers_id': torch.tensor([cid])
  })

  input_inter = input_inter.to(device)
  with torch.no_grad():
      scores = model.full_sort_predict(input_inter.cpu())
  print(scores.shape)
  return (scores)

#call reco
get_recommendation(16128, model)

```

**Comment by laxmimerit** [comment]
**Update 3**
I have also tried to get a recommendation for the user id present in the `.inter` file. But got the same error. Not sure how to make sure that I get the right result. Every time I need to restart run-time if an error is thrown out. Looking for a suggestion to make it stable.

**Comment by chenyushuo** [comment]
Can you offer the full output of error? I'm confused, too.
And you said that it will raise error when you pass some big number of `uid`, can you offer `train_data.dataset.user_num` and `train_data.dataset.item_num`?

**Comment by laxmimerit** [comment]
Just a sec. 

![image](https://user-images.githubusercontent.com/16714612/104592080-a9918680-5693-11eb-9d8b-b27476f74530.png)

Full Error Stack
```
/usr/local/lib/python3.6/dist-packages/IPython/lib/pretty.py in pretty(self, obj)
    396                             if callable(meth):
    397                                 return meth(obj, self, cycle)
--> 398             return _default_pprint(obj, self, cycle)
    399         finally:
    400             self.end_group()

/usr/local/lib/python3.6/dist-packages/IPython/lib/pretty.py in _default_pprint(obj, p, cycle)
    516     if _safe_getattr(klass, '__repr__', None) not in _baseclass_reprs:
    517         # A user-provided repr. Find newlines and replace them with p.break_()
--> 518         _repr_pprint(obj, p, cycle)
    519         return
    520     p.begin_group(1, '<')

/usr/local/lib/python3.6/dist-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    707     """A pprint that just redirects to the normal repr function."""
    708     # Find newlines and replace them with p.break_()
--> 709     output = repr(obj)
    710     for idx,output_line in enumerate(output.splitlines()):
    711         if idx:

/usr/local/lib/python3.6/dist-packages/torch/tensor.py in __repr__(self)
    177             return handle_torch_function(Tensor.__repr__, relevant_args, self)
    178         # All strings are unicode in Python 3.
--> 179         return torch._tensor_str._str(self)
    180 
    181     def backward(self, gradient=None, retain_graph=None, create_graph=False):

/usr/local/lib/python3.6/dist-packages/torch/_tensor_str.py in _str(self)
    370 def _str(self):
    371     with torch.no_grad():
--> 372         return _str_intern(self)

/usr/local/lib/python3.6/dist-packages/torch/_tensor_str.py in _str_intern(self)
    350                     tensor_str = _tensor_str(self.to_dense(), indent)
    351                 else:
--> 352                     tensor_str = _tensor_str(self, indent)
    353 
    354     if self.layout != torch.strided:

/usr/local/lib/python3.6/dist-packages/torch/_tensor_str.py in _tensor_str(self, indent)
    239         return _tensor_str_with_formatter(self, indent, summarize, real_formatter, imag_formatter)
    240     else:
--> 241         formatter = _Formatter(get_summarized_data(self) if summarize else self)
    242         return _tensor_str_with_formatter(self, indent, summarize, formatter)
    243 

/usr/local/lib/python3.6/dist-packages/torch/_tensor_str.py in __init__(self, tensor)
     87 
     88         else:
---> 89             nonzero_finite_vals = torch.masked_select(tensor_view, torch.isfinite(tensor_view) & tensor_view.ne(0))
     90 
     91             if nonzero_finite_vals.numel() == 0:

RuntimeError: CUDA error: device-side assert triggered
```

**Comment by laxmimerit** [comment]
**Update 4:**
Even some of `id's` given by `train_data.dataset.inter_feat.interaction['uid'].sort()` not present in dataset. For example, look at `OUT [61]` in above image for values 
```
torch.return_types.sort(
values=tensor([     1,      1,      1,  ..., 790548, 790549, 790550]),
indices=tensor([ 990497,   22255,  342365,  ..., **1082359**, 1152222, **1351769**]))
```
only values inside `** **` present in the dataset. other indices are not present.

**Comment by laxmimerit** [comment]
**Update 5**
I understood the working principle. Values in the below tensor are actually indices of `uid`
```
torch.return_types.sort(
values=tensor([     1,      1,      1,  ..., 790548, 790549, 790550]),
indices=tensor([ 990497,   22255,  342365,  ..., 1082359, 1152222, 1351769]))
```
**Note: `uid` and `user_id` are the same** I was working with many examples together so got mixed up.

Code was working for lower values because it was being passed as the index value. Code only understands index values [I believe] so to get a recommendation for a particular user, I have to first convert it in the corresponding index. I used `dataset.token2id('uid', [my-actual-user-id])`. It will convert my actual user id into a meaningful id. Again to get actual tokens, I used `dataset.id2token('uid', [1])`.


![image](https://user-images.githubusercontent.com/16714612/104603369-570b9680-56a2-11eb-892b-779262b9e4e8.png)

**How Did I Understand it**
The above figure is from the `ml-100k` dataset. I opened `ml-100k.inter` in excel and sorted `user_id` in ascending order then deleted the first `user_id` data. Thereafter trained it and saw the largest tensor values got reduced to 942 and the lowest kept as it is `1`. The above figure is given for a better understanding for the beginners.
